File: sort/quicksort/java/QuickSort.java
public class QuickSort {
  public static void sort(int[] A) {
    if (A == null) return;
    sort(A, 0, A.length-1);
  }

  private static void sort(int[] A, int low, int high) {
    if (low >= high) return;
    int pivot_pos = partition(A, low, high);
    sort(A, low, pivot_pos-1);
    sort(A, pivot_pos+1, high);
  }

  private static int partition(int[] A, int low, int high) {
    int pivot = A[high];
    int i = low - 1;
    for (int cur = low; cur < high; cur++) {
      if (A[cur] < pivot) {
        swap(A, ++i, cur);
      }
    }
    swap(A, ++i, high);
    return i;
  }

  private static void swap(int[] A, int i, int j) {
    int t = A[i];
    A[i] = A[j];
    A[j] = t;
  }
}

File: sort/quicksort/quick_sort.cpp
#include <algorithm>
#include <vector>
#include "quick_sort.h"

// Forward declare.
int Partition(std::vector<int>& arr, int begin, int end);
void QuicksortRec(std::vector<int>& arr, int begin, int end);

// Quick sort algorithm on integer array.
void Quicksort(std::vector<int>& arr) {
  if (arr.empty()) {
    return;
  }
  int n = arr.size() - 1;
  QuicksortRec(arr, 0, n-1);
}

// Partition array, and return the index of pivot.
int Partition(std::vector<int>& arr, int begin, int end) {
  int pivot = arr[end];
  int i = begin-1;
  int cur = begin;
  while (cur < end) {
    if (arr[cur] < pivot) {
      std::swap(arr[++i], arr[cur]);
    }
    cur++;
  }
  std::swap(arr[++i], arr[cur]);
  return i;
}

// Sort elements in range arr[begin,...,end].
void QuicksortRec(std::vector<int>& arr, int begin, int end) {
  if (begin >= end) {
    return;
  }
  int pivot = Partition(arr, begin, end);
  QuicksortRec(arr, begin, pivot-1);
  QuicksortRec(arr, pivot+1, end);
}
